atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java

```diff
@@ -0,0 +1,182 @@
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.common.base.Throwables;
+
+public final class PersistentLock {
+    public static final long LOCKS_TIMESTAMP = 0L;
+
+    private static final Logger log = LoggerFactory.getLogger(PersistentLock.class);
+
+    private final KeyValueService keyValueService;
+
+    private PersistentLock(KeyValueService keyValueService) {
+        this.keyValueService = keyValueService;
+    }
+
+    public static PersistentLock create(KeyValueService keyValueService) {
+        PersistentLock persistentLock = new PersistentLock(keyValueService);
+        persistentLock.createPersistedLocksTable();
+        return persistentLock;
+    }
+
+    public <T> T runWithExclusiveLock(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason) throws PersistentLockIsTakenException {
+        return runWithLockInternal(supplier, lock, reason, true);
+    }
+
+    public <T> T runWithNonExclusiveLock(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason) throws PersistentLockIsTakenException {
+        return runWithLockInternal(supplier, lock, reason, false);
+    }
+
+    private <T> T runWithLockInternal(
+            Supplier<T> supplier,
+            PersistentLockName lock,
+            String reason,
+            boolean exclusive) throws PersistentLockIsTakenException {
+        LockEntry acquiredLock = acquireLock(lock, reason, exclusive);
+
+        try {
+            return supplier.get();
+        } catch (Exception e) {
+            throw Throwables.throwUncheckedException(e);
+        } finally {
+            releaseLock(acquiredLock);
+        }
+    }
+
+    public LockEntry acquireLock(PersistentLockName lockName, String reason) throws PersistentLockIsTakenException {
+        return acquireLock(lockName, reason, true);
+    }
+
+    public LockEntry acquireLock(
+            PersistentLockName lockName, String reason, boolean exclusive) throws PersistentLockIsTakenException {
+        LockEntry lockEntry = generateUniqueLockEntry(lockName, reason, exclusive);
+
+        insertLockEntry(lockEntry);
+
+        return verifyLockWasAcquired(lockEntry);
+    }
+
+    public void releaseLock(LockEntry lock) {
+        log.debug("Releasing persistent lock {}", lock);
+        keyValueService.delete(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lock.deletionMapWithTimestamp(LOCKS_TIMESTAMP));
```

Comment:
It will not suffer from that issue because each row in this table has a different lockId.